File: lab3/gr1sub1/LANKUT_ALEXANDR/src/pstat.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/sysinfo.h>
#include <ctype.h>
#include <limits.h>
#include <errno.h>

int validate_pid(const char *pid_str) {
    if (!pid_str || !*pid_str) {
        return 0;
    }

    for (const char *p = pid_str; *p; p++) {
        if (!isdigit(*p)) {
            return 0;
        }
    }

    char path[PATH_MAX];
    snprintf(path, sizeof(path), "/proc/%s", pid_str);
    
    if (access(path, F_OK) == -1) {
        return 0;
    }

    return 1;
}

void format_bytes(unsigned long bytes, char *output) {
    if (bytes > 1024 * 1024) {
        snprintf(output, 64, "%.2f MiB", (float)bytes / (1024.0 * 1024.0));
    } else if (bytes > 1024) {
        snprintf(output, 64, "%.2f KiB", (float)bytes / 1024.0);
    } else {
        snprintf(output, 64, "%lu B", bytes);
    }
}

unsigned long read_status_value(FILE *fp, const char *field) {
    char line[PATH_MAX];
    char *value;
    unsigned long result = 0;
    long pos;

    pos = ftell(fp);
    if (pos == -1) {
        return 0;
    }

    while (fgets(line, sizeof(line), fp)) {
        if (strncmp(line, field, strlen(field)) == 0) {
            value = strchr(line, ':');
            if (value) {
                char *endptr;
                errno = 0;
                result = strtoul(value + 1, &endptr, 10);
                if (errno || endptr == value + 1) {
                    result = 0;
                }
                break;
            }
        }
    }

    if (fseek(fp, pos, SEEK_SET) == -1) {
        perror("Failed to reset file position");
    }
    return result;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <pid>\n", argv[0]);
        return 1;
    }

    if (!validate_pid(argv[1])) {
        fprintf(stderr, "Error: Invalid PID or process does not exist\n");
        return 1;
    }

    char path[PATH_MAX];
    char formatted[64];
    FILE *stat_fp, *status_fp, *io_fp;
    char state;
    int ppid, num_threads;
    unsigned long utime, stime;
    unsigned long vctx, nvctx;
    unsigned long vm_rss = 0, rss_anon = 0, rss_file = 0;
    unsigned long read_bytes = 0, write_bytes = 0;
    int have_io_stats = 0;

    snprintf(path, sizeof(path), "/proc/%s/stat", argv[1]);
    stat_fp = fopen(path, "r");
    if (!stat_fp) {
        perror("Failed to open stat file");
        return 1;
    }

    int scan_result = fscanf(stat_fp, "%*d %*s %c %d %*d %*d %*d %*d %*u %*u %*u %*u %*u %lu %lu",
           &state, &ppid, &utime, &stime);
    fclose(stat_fp);

    if (scan_result != 4) {
        fprintf(stderr, "Error reading from stat file (expected 4 values, got %d)\n", scan_result);
        return 1;
    }

    snprintf(path, sizeof(path), "/proc/%s/status", argv[1]);
    status_fp = fopen(path, "r");
    if (!status_fp) {
        perror("Failed to open status file");
        return 1;
    }

    num_threads = read_status_value(status_fp, "Threads");
    vm_rss = read_status_value(status_fp, "VmRSS");
    rss_anon = read_status_value(status_fp, "RssAnon");
    rss_file = read_status_value(status_fp, "RssFile");
    vctx = read_status_value(status_fp, "voluntary_ctxt_switches");
    nvctx = read_status_value(status_fp, "nonvoluntary_ctxt_switches");
    fclose(status_fp);

    snprintf(path, sizeof(path), "/proc/%s/io", argv[1]);
    io_fp = fopen(path, "r");
    if (io_fp) {
        char line[PATH_MAX];
        char *endptr;
        while (fgets(line, sizeof(line), io_fp)) {
            if (strncmp(line, "read_bytes:", 11) == 0) {
                errno = 0;
                read_bytes = strtoul(line + 11, &endptr, 10);
                if (errno || endptr == line + 11) {
                    read_bytes = 0;
                }
            } else if (strncmp(line, "write_bytes:", 12) == 0) {
                errno = 0;
                write_bytes = strtoul(line + 12, &endptr, 10);
                if (errno || endptr == line + 12) {
                    write_bytes = 0;
                }
            }
        }
        fclose(io_fp);
        have_io_stats = 1;
    }

    printf("\nProcess %s information:\n", argv[1]);
    printf("---------------------------\n");
    printf("Parent PID: %d\n", ppid);
    printf("Thread count: %d\n", num_threads);
    printf("State: %c", state);
    switch(state) {
        case 'R': printf(" (running)\n"); break;
        case 'S': printf(" (sleeping)\n"); break;
        case 'D': printf(" (disk sleep)\n"); break;
        case 'Z': printf(" (zombie)\n"); break;
        case 'T': printf(" (stopped)\n"); break;
        default: printf("\n");
    }

    long clock_ticks = sysconf(_SC_CLK_TCK);
    if (clock_ticks <= 0) {
        fprintf(stderr, "Warning: Could not determine clock ticks per second\n");
        clock_ticks = 100;  // Используем типичное значение как запасной вариант
    }
    double cpu_time = (double)(utime + stime) / clock_ticks;
    printf("CPU Time: %.2f sec (user: %lu, system: %lu)\n", 
           cpu_time, utime, stime);

    printf("Context switches: voluntary=%lu, nonvoluntary=%lu\n", 
           vctx, nvctx);

    format_bytes(vm_rss * 1024, formatted);
    printf("Memory usage (VmRSS): %s\n", formatted);
    format_bytes(rss_anon * 1024, formatted);
    printf("Anonymous memory (RssAnon): %s\n", formatted);
    format_bytes(rss_file * 1024, formatted);
    printf("File cache (RssFile): %s\n", formatted);

    if (have_io_stats) {
        format_bytes(read_bytes, formatted);
        printf("Read: %s\n", formatted);
        format_bytes(write_bytes, formatted);
        printf("Written: %s\n", formatted);
    }
    printf("---------------------------\n");

    return 0;
}

File: lab3/lab3/gr9sub2/Kedank_Valeriya/src/pstat.py
import sys
import os

def print_human_readable_kb(kb):
    if kb > 1024:
        print(f"{kb / 1024:.2f} MiB", end="")
    else:
        print(f"{kb} KiB", end="")

def read_stat(pid):
    path = f"/proc/{pid}/stat"
    try:
        with open(path) as f:
            parts = f.read().split()
            comm = parts[1].strip("()")
            state = parts[2]
            ppid = int(parts[3])
            utime = int(parts[13])
            stime = int(parts[14])
            hz = os.sysconf(os.sysconf_names['SC_CLK_TCK'])
            cpu_time = (utime + stime) / hz
            return comm, state, ppid, cpu_time
    except Exception as e:
        print(f"[ERROR] stat: {e}")
        sys.exit(1)

def read_status(pid):
    path = f"/proc/{pid}/status"
    data = {
        "Threads": 0,
        "VmRSS": 0,
        "RssAnon": 0,
        "RssFile": 0,
        "voluntary_ctxt_switches": 0,
        "nonvoluntary_ctxt_switches": 0
    }
    try:
        with open(path) as f:
            for line in f:
                for key in data:
                    if line.startswith(key + ":"):
                        value = int(line.split()[1])
                        data[key] = value
        return data
    except Exception as e:
        print(f"[ERROR] status: {e}")
        sys.exit(1)

def read_io(pid):
    path = f"/proc/{pid}/io"
    read_bytes = write_bytes = 0
    try:
        with open(path) as f:
            for line in f:
                if line.startswith("read_bytes:"):
                    read_bytes = int(line.split()[1])
                elif line.startswith("write_bytes:"):
                    write_bytes = int(line.split()[1])
        return read_bytes, write_bytes
    except Exception as e:
        print(f"[ERROR] io: {e}")
        sys.exit(1)

def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <pid>")
        sys.exit(1)

    pid = int(sys.argv[1])
    comm, state, ppid, cpu_time = read_stat(pid)
    status = read_status(pid)
    read_bytes, write_bytes = read_io(pid)

    print(f"Process {pid} ({comm})")
    print(f"  PPid: {ppid}")
    print(f"  State: {state}")
    print(f"  Threads: {status['Threads']}")
    print(f"  CPU time: {cpu_time:.2f} sec (utime+stime)")
    print(f"  Context switches: voluntary={status['voluntary_ctxt_switches']}, nonvoluntary={status['nonvoluntary_ctxt_switches']}")

    print("  Memory:")
    print("    VmRSS: ", end=""); print_human_readable_kb(status['VmRSS']); print()
    print("    RssAnon: ", end=""); print_human_readable_kb(status['RssAnon']); print()
    print("    RssFile: ", end=""); print_human_readable_kb(status['RssFile']); print()

    print("  IO:")
    print(f"    read_bytes: {read_bytes}")
    print(f"    write_bytes: {write_bytes}")

if __name__ == "__main__":
    main()

